apps/site/assets/ts/stop/components/DepartureTimes.tsx

```diff
@@ -1,74 +1,299 @@
 import React, { ReactElement } from "react";
 import usePredictionsChannel from "../../hooks/usePredictionsChannel";
-import { isACommuterRailRoute } from "../../models/route";
 import { DirectionId, Route, Stop } from "../../__v3api";
 import renderFa from "../../helpers/render-fa";
-import { formatDepartureTime } from "../../helpers/date";
+import { formatToBostonTime } from "../../helpers/date";
+import { find, groupBy, slice, some } from "lodash";
+import {
+  differenceInSeconds,
+  isSameDay,
+  secondsInDay,
+  secondsInHour,
+  secondsInMinute
+} from "date-fns";
+import realtimeIcon from "../../../static/images/icon-realtime-tracking.svg";
+import SVGIcon from "../../helpers/render-svg";
+import { ScheduleWithTimestamp } from "../../models/schedules";
+import { PredictionWithTimestamp } from "../../models/perdictions";
+import {
+  departureInfoToTime,
+  mergeIntoDepartureInfo
+} from "../../helpers/departureInfo";
+import { DepartureInfo } from "../../models/departureInfo";
 
-interface DepartureTimesProps {
-  route: Route;
-  stop: Stop;
-  directionId: DirectionId;
+// This interface is used to tell the front end
+// how to display the ScheduleInfoModel data
+interface DisplayTimeConfig {
+  isPrediction?: boolean;
+  displayString: string;
+  trackName?: string | null;
+  isTomorrow?: boolean;
+  isBolded?: boolean;
+  isStikethrough?: boolean;
 }
 
+const getNextUnCancelledDeparture = (
+  departureInfos: DepartureInfo[]
+): DepartureInfo | undefined => {
+  return find(departureInfos, (di: DepartureInfo) => !di.isCancelled);
+};
+
+// This function returns time info in a consisten format.
+// Returns the time from the predictions if they exist, or the schdule
+const getNextTwoTimes = (
+  departureInfos: DepartureInfo[]
+): [boolean, DepartureInfo | undefined, DepartureInfo | undefined] => {
+  const departure1 = departureInfos[0];
```

Comment:
The way cancelled times work is that only the first time in `departureInfos` could be cancelled.  If any other time in the `departureInfos` is cancelled, it will be filtered out by `getNextUncancelledDeparture`.  If the first time is not cancelled, then the other returned time will also not be cancelled.

(I ended up removing the `areAnyPredictions` check.)
So the return from `getNextTwoTimes` is `[time, time]`, `[time, undefined]`, `[cancelled_time, time]`, `[cancelled_time, undefined]` or `[undefined, undefined]`

Having the first departureInfo be cancelled is exactly how this function works.
The showing and deciding display order happens in the `infoToDisplayTime`